### DISCUSSION TITLE
APK export - WARNING: could not find appropriate Java Home location for v11+ JDK

### DISCUSSION THREAD

**Discussion by Tiuser4567** [airsdk/Adobe-Runtime-Support#2198] [Q&A] [answered]
Hello,

We updated to the latest AIR SDK 33.1.1.929 (from 33.1.1.575) and are still exporting as APK via adt.jar  (we don't want to switch to app bundle atm and google playstore seems to still allow APK uploads for existing apps and only requires app bundles for new apps)

However, we found that after updating the build script is showing some new warning
 [java] WARNING: could not find appropriate Java Home location for v11+ JDK
though the command still succeeded and the APK files are still generated and can be installed and seems to run ok on devices.

From my initial check on other discussions on similar titles, it seems above warning is more likely related to the generation of app bundle ? (which requires the JAVA_HOME setting in adt.cfg) but I couldn't find any similar discussion related to APKs.

1. Then does this mean we can just ignore above warning as we are still exporting as APK and not app bundle? 
(currently we are just using JDK 8 though and not JDK11 so can we still use JDK 8 as long we are not exporting as app bundle)
2. As a followup question, if we want to export as app bundle in the future, do we need to update to use JDK 11+ ?

**Comment by ajwfrost** [comment, marked as answer]
Hi

We now try to build both APK and AAB files via the Gradle mechanism, the same as normal Android apps built with Android Studio, so that's probably where the warning is coming from. If we're unable to find the JDK 11+ then we'd fall back to the 'legacy' mechanism (which you can also specify in your app descriptor file if you wish, using a setting within the `<android>` section of `<BuildLegacyAPK>true</BuildLegacyAPK>`)

The only issue here is if you're using ANEs, the latest versions of these would have also been updated to use the Gradle-based mechanism so the legacy mechanism may cause problems if you've updated e.g. to the latest version of Distriqt extensions.

So if you do encounter weird issues with the APKs that you're creating using Java 8 or the legacy build method, then it would be worth getting Android Studio and the JDK that comes with this which we should be able to detect and use... otherwise, yes you can ignore the warning.

For app bundles, yes the Android toolchain now requires JDK 11+. Basically we're aiming to roughly match the versions used in Android Studio which will itself provide the JDK, so we try to use that if it's installed in a standard location.

thanks

**Comment by teotigraphix** [comment]
> The only issue here is if you're using ANEs, the latest versions of these would have also been updated to use the Gradle-based mechanism so the legacy mechanism may cause problems if you've updated e.g. to the latest version of Distriqt extensions.

@ajwfrost I have an app that runs fine launching from IDEA but when I try to install the apk 32/64 doesn't matter, all I get is a black screen. Never had this before, so are you saying that the ANE MUST be recompiled with this gradle system? I have my own ane that I haven't recompiled in ages.

**EDIT**: Now that I think about it, when I have been scanning the Logcat messages, it does seem to stop when loading my ane. Will recompile and see.

![image](https://user-images.githubusercontent.com/226417/195662883-6cdb1739-80a4-450d-a35f-bb467f473c96.png)

**Reply by teotigraphix** [reply]
I recompiled my ANE and absolutely no difference, just a black screen. I have the APK @ajwfrost if you would like to try and load it, debug it.

**Reply by ajwfrost** [reply]
Hi

Looking at that logcat and the messages from the same process/thread it looks like ANE is doing something with hardware graphics (via mapper 4.x, 3.x) which isn't working ... if you don't see further messages tagged with this process/thread then it seems to me as if something in the ANE has hung it... you can see the "AIR VERSION" output at the top, there should then be a similar output (I think using commas for the version number) a bit later once the bootstrap has finished and the initial SWF is loaded.

The changes made in the ANE side aren't about just recompiling, but are more about custom resources. So if you had any resources listed in packages via the platform options XML file, the namespaces for these should be corrected so that they are properly read. If you've not adjusted your ANE then you should probably stick with the "legacy" build method. I'm not actually sure which method you're using when doing the packaging here.... (if you get the warning about "captive runtime" then it's the legacy one, we took that out for the Gradle builds..)

One other thing comes to mind, in case your ANE is also trying to access the original AIR activity .. the name for this has changed, from "AppEntry" to "AIRAppEntry" I think it is. So worth checking for that too... if you're able to run this via Android Studio and debug your ANE code, it may help..?

thanks

**Reply by teotigraphix** [reply]
Well none of those listed items are on the true side you listed as far as the ANE, has no resources, nothing to do with the Activity, etc.

I have to do some more research before I say anything else. I need to find where it stopped working with SDKs. I have been testing this on the desktop mainly so I would not have seen where trying to "install" an apk would have failed. This ALWAYS runs from IDEA. So does that give you any other clues.. ? The Debug/Run configs run this and the ANE loads fine on both a Pixel and an old Nexus 7 2013.

**Reply by ajwfrost** [reply]
If it runs - as a remote process on the device? - via IDEA, then I'd be interested to know how they're building and deploying it.. could you try enabling the debug output for ADT and we can see what's happening in more detail? (Find the "adt.cfg" file in the AIR SDK's "lib" folder, and edit the line to remove the `#` comment character to enable `DebugOut=true`. Then after running it, please check the file c:\users\username\adt.log).

thanks

**Reply by teotigraphix** [reply]
Cool Thanks @ajwfrost I have some things to do today but this is priority and will get back to you, thank you for your time.

> If it runs - as a remote process on the device? 

By this I am speaking of, creating a run/debug config in IDEA and hitting Shift+F9, IDEA compiles, then runs on the device from USB debugging. I have never in this whole time had a problem running my audio engine on Android through the ANE. 

The quick backstory... I dev with windows, but am ready to "release" so before I hit the Play test track I wanted to generate apks for some testers I know. So I have a ANT build script that either builds an APK or AAB. So as of 2 days ago I started to try and install the built APK with the ANT script, no go, thus this comment.

**Reply by ajwfrost** [reply]
Thanks - the other thing you could do is if you build with "apk-debug" (or "aab-debug") then send us the generated file, we can install it and (on older phones) replace the embedded AIR runtime library with our own one which means we can do more debugging..

Plus perhaps just to check the environment you could try building an app that is blank/empty? or just take out the ANE to see if that's the issue or if there's something else going on (not sure how far it would get ...)

Thinking again about the screenshot of the log above .. when that call to the ANE happens, I don't think it should actually be doing anything other than registering it and loading the SWF definitions. So - unless you have static/class properties that would trigger the native library to be loaded - it shouldn't do anything within the native library itself until the extension context is created... Is it a C++ or Java library?

thanks

**Reply by teotigraphix** [reply]
> build with "apk-debug" (or "aab-debug") then send us the generated file

I will do this as well.

> try building an app that is blank/empty

Yup, was on my list if you hadn't said anything obvious, which is where we are at right now.

>  Is it a C++ or Java library?

The engine is C++ with a C interface(to call) that then is wrapped with the jar FRE implementations.

The audio engine is bootstrapped(my logs on the device form IDEA debug USB);

```
"C:\Program Files\JetBrains\IntelliJ IDEA 2022.2.2\jbr\bin\java.exe" -Dapplication.home=E:\Dev\AIR\SDK\AIR-SDK-Windows-33.1.1.929 -Dfile.encoding=UTF-8 -Djava.awt.headless=true -Duser.language=en -Duser.region=en -Xmx512m -classpath E:/Dev/AIR/SDK/AIR-SDK-Windows-33.1.1.929/lib/legacy/fdb.jar flex.tools.debugger.cli.DebugCLI -p 7936
Adobe fdb (Flash Player Debugger) [build development]
Copyright (c) 2004-2007 Adobe, Inc. All rights reserved.
Trying to connect to Player
Player connected; session starting.
[trace] Main
[trace] [Starling] Context ready. Display Driver: OpenGL (Enhanced)
[trace] [AssetManager] Enqueuing 'Theme.xml'
[trace] [AssetManager] Enqueuing 'Theme.png'
[trace] TMXTheme new TMXTheme()
[trace] [AssetManager] Adding xml 'Theme'
[trace] [AssetManager] Adding texture 'Theme'
[trace] [AssetManager] Adding textureAtlas 'Theme'
[trace] ToneMatrix new ToneMatrix() [object Stage]
[trace] AriesDrawersApplication this_addedToStageHandler(starling.events.Event)
[trace] AriesStartup boot() 13 [object TMXStartup] null
[trace] AriesStartup Configure Application - Aries.VERSION SDK Version 0.0.1
[trace]        [Log]                                  ( AriesContext )    13     new AriesContext()
[trace]        [Log]                              ( AriesApplication )    13     new AriesApplication()
[trace] RuntimeUtils Configure File System
[trace] RuntimeUtils     RuntimeUtils.rootDirectory app-storage:/
[trace] RuntimeUtils     RuntimeUtils.applicationDirectory app-storage:/ToneMatrix
[trace]        [Log]                                  ( AriesStartup )    13 stageWidth 1080 stageHeight 1920
[trace]        [Log]                                  ( AriesStartup )    13 isPhone true
[trace]        [Log]                                  ( AriesStartup )    13 isLargePhone true
[trace]        [Log]                                  ( AriesStartup )    13 isTablet false
[trace]        [Log]                                  ( AriesStartup )    13 dpi 420
[trace]        [Log]                                  ( AriesStartup )    13 screenPixelWidth NaN screenPixelHeight NaN
[trace]        [Log]                                  ( AriesStartup )    13 fullScreenWidth 1080 fullScreenHeight 1920
[trace] AbstractApplicationBootstrap END starling_rootCreatedHandler()
[trace] AbstractApplicationBootstrap ============================================================
[trace] AbstractApplicationBootstrap
[trace]      [Debug]                                  ( AriesStartup )    15 startAndRun(15) >>>>
[trace]      [Debug]                                  ( AriesStartup )    15 initialize CausticCoreImpl
[trace]      [Debug]                                  ( AriesStartup )    15 new CausticCoreImpl()
[trace]      [Debug]                                  ( AriesStartup )    15 Create CausticCoreImpl
```

This is the first call into the native ANE, I WILL double check but it has to be, I have worked on this a long time. ;-)

So I just ran the ToneMatrix app on the Pixel 2 with Android 11 installed and no problems. But creating the apk and installing.. the problem.

`CausticCoreImpl` is the ANE impl in AS3.
